include/kodlab_mjbots_sdk/behavior.h

```diff
@@ -0,0 +1,205 @@
+   *          appropriate variables to begin the behavior. The previously active
+   *          behavior is passed as an argument, permitting different
+   *          functionalities for different prior behaviors.
+   * @param prev_behavior previously active behavior
+   */
+  virtual void Begin(const Behavior<Robot> &prev_behavior) {}
+
+  /**
+   * @brief Update this behavior for a given robot interface
+   * @note This function must be implemented in a derived class
+   * @details This code runs after Begin, and is called on every control loop
+   *          update. Use this to define the primary content of your behavior,
+   *          e.g., to update and command new joint torques.  Additionally, this
+   *          code can monitor if the behavior has terminated and update flags
+   *          appropriately.
+   * @param robot robot interface
+   */
+  virtual void Update() = 0;
+
+  /**
+   * @brief Stop this behavior
+   * @details This code is run when a behavior is directed to stop running by
+   *          the behavior manager. This should trigger a shutdown process for
+   *          the behavior, e.g. slowing the robot motion. The upcoming behavior
+   *          is passed as an argument to permit different shutdown procedures
+   *          for different on-deck behaviors.
+   * @param next_behavior behavior to be transitioned to
+   */
+  virtual void Stop(const Behavior<Robot> &next_behavior) {}
+
+  /**
+   * @brief Check for whether the behavior is running
+   * @details This function identifies if the behavior is active. By default, it
+   *          simply accesses the `active_` variable which is toggled by the
+   *          behavior manager.
+   * @return true if this behavior is active, false otherwise
+   */
+  virtual bool Running() { return active_; }
+
+  /**
+   * @brief Check for whether this behavior is prepared to switch to a new
+   *        behavior
+   * @details This function indicates to the behavior manager whether the
+   *          behavior is prepared to terminate and transition to the next
+   *          behavior. In its simplest form, this function accesses the
+   *          internal behavior active flag. This function can be used as a mode
+   *          switch guard.
+   * @param next_behavior behavior being transitioned to
+   * @return true if behavior is ready to switch, false otherwise
+   */
+  virtual bool ReadyToSwitch(const Behavior<Robot> &next_behavior)
+  {
+    return true;
+  }
+
+  /**
+   * @brief Processes input data, if available
+   * @details This method should be implemented by child classes wishing to
+   * handle input processing internally. For supported LCM behavior inputs,
+   * please inherit `IOBehavior`.
+   * @note If using a behavior manager, this method is called every update
+   * cycle, and should implement checking to determine if a new input is
+   * available prior to using the data.
+   */
+  virtual void ProcessInput() {}
```

Comment:
I'm a little confused why you have these functions here even though this is not an IO behavior.

Reply:
reading forward I don't think these should be here.

Reply:
This is here because the system currently uses the a vector of pointers to behaviors (upcasted), there are more complicated ways to deal with this issue.

Reply:
Yes as @jdcaporale said, this was the path of least resistance to allow the `BehaviorManager` to maintain a vector of pointers to `Behavior`-derived classes.  

`Behavior` contains all of the definitions for methods that may be called by `BehaviorManager`.  To allow `ProcessInput` and `ProcessOutput` to be called by `BehaviorManager` when they are used in `IOBehavior`, these definitions must also exist here.  

Alternative solutions could be implemented on the `BehaviorManager` side of things:

- `BehaviorManager` could attempts to downcast the active `Behavior` pointer to an `IOBehavior`, and then only call `ProcessInput` and `ProcessOutput` if the downcast was successful.
- An `AbstractBehavior` class could be made which includes all of the necessary method definitions and is a parent to both `Behavior` and `IOBehavior`.  Then, the vector in `BehaviorManager` could hold pointers to `AbstractBehavior`s instead.

Reply:
ok. I think thats reasonable.